### PR TITLE
LPS-199218 Create async & headles item actions

### PR DESCRIPTION
**Pull request by juanjofgliferay**
## Details
Related story/task: [LPS-191485](https://liferay.atlassian.net/browse/LPS-191485) / [LPS-199218](https://liferay.atlassian.net/browse/LPS-199218)
Figma: [Item actions designs](https://www.figma.com/file/87FEbztFRRf0OqJJhiqT0U/LPS-186881-configure-actions-for-dataset-views?type=design&node-id=224-44674&mode=design&t=shDvRzbSZ6gWBWHe-0)

### Changes from original design
- Headless type doesn't have the method picker/selector [see slack thread](https://liferay.slack.com/archives/C3JBR21HA/p1698057700420559?thread_ts=1697207711.213759&cid=C3JBR21HA)

## Goal
Admin users could define item actions of type `async` and `headless`

Needs `Feature Flag: LPS-194395`

## UI
[async-headless-actions.webm](https://github.com/liferay-frontend/liferay-portal/assets/132653342/5a90ce32-0190-41af-8981-0c223c44f984)

### CODE REVIEW

**Comment by juanjofgliferay** [comment]
ci:test:sf

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/css/Actions.scss [hunk 1]
```diff
@@ -14,4 +14,12 @@
...
 		}
 	}
+
+	.tab-content {
+		&.action-messages {
+			.tab-pane {
+				padding: 2rem 0;
+			}
+		}
+	}
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/css/Actions.scss:24]: We can simplify this to:
```js
.tab-content.action-messages .tab-pane {
	padding: 2rem 0;
}
```

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/actions/ActionForm.tsx [hunk 2]
```diff
@@ -21,6 +22,13 @@ import openDefaultFailureToast from '../../utils/openDefaultFailureToast';
 import openDefaultSuccessToast from '../../utils/openDefaultSuccessToast';
 import {IFDSAction} from '../Actions';
 
+const ACTION_METHOD = {
+	DELETE: 'DELETE',
+	GET: 'GET',
+	PATCH: 'PATCH',
+	POST: 'POST',
+};
+
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/actions/ActionForm.tsx:31]: (optional) You can consider `enum`s for this type of usage, which may be more descriptive than simple objects. We started adding them in [few places in DSM](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/types.ts#L8-L18). It's optional, simple objects is fine as well.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/actions/ActionForm.tsx [hunk 3]
```diff
@@ -107,18 +126,40 @@ const ActionForm = ({
...
 }: IFDSActionFormProps) => {
+	const ACTION_METHODS = useMemo(() => {
+		const methods = [];
+
+		for (const method in ACTION_METHOD) {
+			methods.push({label: method, value: method});
+		}
+
+		return methods;
+	}, []);
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/actions/ActionForm.tsx:137]: Is there a reason why this is in the component? It seems to me that this should be a top level constant, sibling to `ACTION_TYPES`. With that placement, there would be no need for `useMemo`.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/actions/ActionForm.tsx [hunk 4]
```diff
@@ -390,9 +452,15 @@ const ActionForm = ({
...
 									}
 									options={
-										Liferay.FeatureFlags['LPS-194395']
-											? ACTION_TYPES
-											: ACTION_TYPES.slice(0, 1)
+										activeTab === 0
+											? Liferay.FeatureFlags['LPS-194395']
+												? ITEM_ACTION_TYPES.concat(
+														ACTION_TYPES
+												  )
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/actions/ActionForm.tsx:459]: We can avoid code duplication here, by extracting concatenation to definition. Something like:
```js
const ITEM_ACTION_TYPES = [
	...
].concat(ACTION_TYPES);
```

**Review by markocikos — CHANGES_REQUESTED** [review]
@juanjofgliferay Please see comments inline. In addition, there is something weird with stack of commits in this PR, there are a bunch of modified `LPS-197771` commits, that were suppose to be merged. Please cherry pick or create new commits on top of `master`.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/actions/ActionForm.tsx (hunk 3 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/actions/ActionForm.tsx:137]: I was not sure about the usage of the `useMemo` outside of the React component and basically moving it to a constant sibling to `ACTION_TYPES` should be enough.

**Comment by juanjofgliferay** [comment]
Follow up on https://github.com/liferay-frontend/liferay-portal/pull/3769

**Pull request closed without merging**